upi/openstack/security-groups.yaml

```diff
@@ -199,155 +221,171 @@
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: udp
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
       port_range_min: 4500
       port_range_max: 4500
 
   - name: 'Create worker-sg rule "worker ingress internal (TCP)"'
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: tcp
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
       port_range_min: 9000
       port_range_max: 9999
 
   - name: 'Create worker-sg rule "worker ingress internal (UDP)"'
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: udp
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
       port_range_min: 9000
       port_range_max: 9999
 
   - name: 'Create worker-sg rule "worker ingress kubelet insecure"'
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: tcp
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
       port_range_min: 10250
       port_range_max: 10250
 
   - name: 'Create worker-sg rule "worker ingress services (TCP)"'
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: tcp
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
       port_range_min: 30000
       port_range_max: 32767
 
   - name: 'Create worker-sg rule "worker ingress services (UDP)"'
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: udp
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
       port_range_min: 30000
       port_range_max: 32767
 
   - name: 'Create worker-sg rule "VRRP"'
     openstack.cloud.security_group_rule:
       security_group: "{{ os_sg_worker }}"
+      ethertype: "{{ 'IPv6' if ipv6_primary else 'IPv4' }}"
       protocol: '112'
-      remote_ip_prefix: "{{ os_subnet_range }}"
+      remote_ip_prefix: "{{ os_subnet6_range if ipv6_primary else os_subnet_range }}"
 
-  - name: 'Create security groups for IPv6'
+  - name: "Create security groups for {{ 'IPv6' if ipv6_primary else 'IPv4' }}"
```

Comment:
It's confusing that we're setting `IPv4` as ethertype when using `IPv6` primary inside this block.
Looks like there are various rules that don't need to be touched, like `master ingress HTTP (TCP)`.  We were already enabling IPv4 and IPv6 at the same time without the change proposed. Can you revisit the rules and not update the ones that already enable IPv4 and IPv6?
Here are the IPv6 rules that were included when adding IPv6 primary support on IPI https://github.com/openshift/installer/commit/2748047e75b2644fb167110b5b9c79a28ede2441

Reply:
Maybe it's worth not changing the rules outside this block, and just add a few more rules that are missing for the IPv6 primary scenario?